build-system-abstraction.rst

```diff
@@ -0,0 +1,472 @@
+the necessary data in a METADATA file. PEP-426 can't be used as it's still
+draft, and defining a new metadata format, while we should do that, is a
+separate problem. Using a directory on disk would not add any value to the
+interface (pip has to do that today due to limitations in the setuptools
+CLI).
+
+The use of 'develop' as a command is because there is no PEP specifying the
+interoperability of things that do what 'setuptools develop' does - so we'll
+need to define that before pip can take on the responsibility for doing the
+'develop' step. Once that's done we can issue a successor PEP to this one.
+
+The use of a command line API rather than a Python API is a little
+contentious. Fundamentally anything can be made to work, and the pip
+maintainers have spoken strongly in favour of retaining a process based
+interface - something that is mature and robust in pip today.
+
+The choice of JSON as a file format is a compromise between several
+constraints. Firstly there is no stdlib YAML interpreter, nor one for any of
+the other low-friction structured file formats. Secondly, INIParser is a poor
+format for a number of reasons, primarily that it has very minimal structure -
+but pip's maintainers are not fond of it. JSON is in the stdlib, has
+sufficient structure to permit embedding anything we want in future without
+requiring embedded DSL's.
+
+Donald suggested using ``setup.cfg`` and the existing setuptools command line
+rather than inventing something new. While that would permit interoperability
+with less visible changes, it requires nearly as much engineering on the pip
+side - looking for the new key in setup.cfg, implementing the non-installed
+environments to run the build in. And the desire from other build system
+authors not to confuse their users by delivering something that looks like but
+behaves quite differently to setuptools seems like a bigger issue than pip
+learning how to invoke a custom build tool.
+
+The metadata and wheel commands are required to have consistent metadata to
+avoid a race condition that could otherwise happen where pip reads the
+metadata, acts on it, and then the resulting wheel has incompatible
+requirements. That race is exploited today by packages using PEP-426
+environment markers, to work with older pip versions that do not support
+environment markers. That exploit is not needed with this PEP, because either
+the setuptools shim is in use (with older pip versions), or an environment
+marker ready pip is in use. The setuptools shim can take care of exploiting
+the difference older pip versions require.
+
+We discussed having an sdist verb. The main driver for this was to make sure
+that build systems were able to produce sdists that pip can build - but this is
+circular: the whole point of this PEP is to let pip consume such sdists
+reliably and without requiring an implementation of setuptools. Further, while
+most everyone agrees that encouraging sdists to be uploaded to PyPI, there
```

Comment:
typo: ``encouraging ... "is good"`` or ``uploading sdists to PyPI is encouraged``.